kinesis/src/main/scala/akka/stream/alpakka/kinesis/scaladsl/KinesisSchedulerSource.scala

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.kinesis.scaladsl
+
+import java.util.concurrent.Semaphore
+
+import akka.stream.Supervision.{Resume, Stop}
+import akka.stream._
+import akka.stream.scaladsl.{Flow, GraphDSL, Keep, Sink, Source, Zip}
+import akka.{Done, NotUsed}
+import akka.stream.alpakka.kinesis.Errors.{BackpressureTimeout, SchedulerUnexpectedShutdown}
+import akka.stream.alpakka.kinesis.{
+  CommittableRecord,
+  KinesisSchedulerCheckpointSettings,
+  KinesisSchedulerSourceSettings,
+  ShardProcessor
+}
+import software.amazon.kinesis.coordinator.Scheduler
+import software.amazon.kinesis.exceptions.ShutdownException
+import software.amazon.kinesis.processor.{ShardRecordProcessor, ShardRecordProcessorFactory}
+import software.amazon.kinesis.retrieval.KinesisClientRecord
+
+import scala.collection.immutable
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.control.Exception
+import scala.util.{Failure, Success}
+
+object KinesisSchedulerSource {
+
+  def apply(
+      schedulerBuilder: ShardRecordProcessorFactory => Scheduler,
+      settings: KinesisSchedulerSourceSettings = KinesisSchedulerSourceSettings.defaultInstance
+  )(implicit schedulerExecutor: ExecutionContext): Source[CommittableRecord, Scheduler] =
+    Source
+      .queue[CommittableRecord](settings.bufferSize, OverflowStrategy.backpressure)
+      .watchTermination()(Keep.both)
+      .mapMaterializedValue {
+        case (queue, watch) =>
+          val semaphore = new Semaphore(1, true)
```

Comment:
Wow, haven't seen real blocking for a while, can this be solved in a non-blocking way?

Reply:
Probably, but I have not found a better and/or simpler alternative than this implementation.

The thing is that the `Scheduler` lives in its own thread (it doesn't return control to the calling thread once you invoke `run()`) and we need to transmit backpressure to it, so I think that the callback needs to block somehow.